### ISSUE TITLE
BuildPhysicsWorld dependency

### ISSUE THREAD

**Issue by Xroft666** [reeseschultz/ReeseUnityDemos#65]
There is something odd with the way your systems are updated relative to the physics system.
I've imported your package as a third party, and every time there is a raycasting used or any read from physicsWorld, I get errors spamming like this

```
InvalidOperationException: The previously scheduled job NavLerpSystem:NavWalkJob reads from the Unity.Collections.NativeHashMap`2[Unity.Entities.Entity,System.Int32] NavWalkJob.JobData.physicsWorld.CollisionWorld.EntityBodyIndexMap. You must call JobHandle.Complete() on the job NavLerpSystem:NavWalkJob, before you can write to the Unity.Collections.NativeHashMap`2[Unity.Entities.Entity,System.Int32] safely.
Unity.Collections.LowLevel.Unsafe.AtomicSafetyHandle.CheckWriteAndBumpSecondaryVersion (Unity.Collections.LowLevel.Unsafe.AtomicSafetyHandle handle) (at <cf2386df99904f30a4c5e83b165f15d0>:0)
Unity.Collections.NativeHashMap`2[TKey,TValue].CheckWrite () (at Library/PackageCache/com.unity.collections@0.15.0-preview.21/Unity.Collections/NativeHashMap.cs:554)
Unity.Collections.NativeHashMap`2[TKey,TValue].Clear () (at Library/PackageCache/com.unity.collections@0.15.0-preview.21/Unity.Collections/NativeHashMap.cs:194)
Unity.Physics.CollisionWorld.Reset (System.Int32 numStaticBodies, System.Int32 numDynamicBodies) (at Library/PackageCache/com.unity.physics@0.6.0-preview.3/Unity.Physics/Collision/World/CollisionWorld.cs:53)
Unity.Physics.PhysicsWorld.Reset (System.Int32 numStaticBodies, System.Int32 numDynamicBodies, System.Int32 numJoints) (at Library/PackageCache/com.unity.physics@0.6.0-preview.3/Unity.Physics/Dynamics/World/PhysicsWorld.cs:41)
Unity.Physics.Systems.BuildPhysicsWorld.OnUpdate () (at Library/PackageCache/com.unity.physics@0.6.0-preview.3/Unity.Physics/ECS/Base/Systems/BuildPhysicsWorld.cs:122)
Unity.Entities.SystemBase.Update () (at Library/PackageCache/com.unity.entities@0.17.0-preview.41/Unity.Entities/SystemBase.cs:412)
Unity.Entities.ComponentSystemGroup.UpdateAllSystems () (at Library/PackageCache/com.unity.entities@0.17.0-preview.41/Unity.Entities/ComponentSystemGroup.cs:472)
UnityEngine.Debug:LogException(Exception)
Unity.Debug:LogException(Exception) (at Library/PackageCache/com.unity.entities@0.17.0-preview.41/Unity.Entities/Stubs/Unity/Debug.cs:19)
Unity.Entities.ComponentSystemGroup:UpdateAllSystems() (at Library/PackageCache/com.unity.entities@0.17.0-preview.41/Unity.Entities/ComponentSystemGroup.cs:477)
Unity.Entities.ComponentSystemGroup:OnUpdate() (at Library/PackageCache/com.unity.entities@0.17.0-preview.41/Unity.Entities/ComponentSystemGroup.cs:423)
Unity.Entities.ComponentSystem:Update() (at Library/PackageCache/com.unity.entities@0.17.0-preview.41/Unity.Entities/ComponentSystem.cs:114)
Unity.Entities.ComponentSystemGroup:UpdateAllSystems() (at Library/PackageCache/com.unity.entities@0.17.0-preview.41/Unity.Entities/ComponentSystemGroup.cs:472)
Unity.Entities.ComponentSystemGroup:OnUpdate() (at Library/PackageCache/com.unity.entities@0.17.0-preview.41/Unity.Entities/ComponentSystemGroup.cs:417)
Unity.Entities.ComponentSystem:Update() (at Library/PackageCache/com.unity.entities@0.17.0-preview.41/Unity.Entities/ComponentSystem.cs:114)
Unity.Entities.DummyDelegateWrapper:TriggerUpdate() (at Library/PackageCache/com.unity.entities@0.17.0-preview.41/Unity.Entities/ScriptBehaviourUpdateOrder.cs:333)
```

To the moment, I am fixing this by adding Dependency.Complete(); after scheduling

**Comment by reeseschultz** [comment]
@Xroft666 this isn't an issue with the nav package itself. The `NavLerpSystem` raycasts under certain conditions for the benefit of obstacle avoidance, so there's nothing odd about it. That's just how it works.

My recommendation is that you try using `UpdateAfter(typeof(NavLerpSystem))` or `UpdateBefore(typeof(NavLerpSystem))`. To understand why, please read [this documentation on system update order](https://docs.unity3d.com/Packages/com.unity.entities@0.17/manual/system_update_order.html?q=system%20update%20order). By "fixing this" with `Dependency.Complete()`, you may be inadvertently introducing an unwanted synchronization point.

**Comment by Xroft666** [comment]
Yeah, I get it. What I am pointing out at though, it is not any of my systems that throw the error.
It is Unity.Physics.PhysicsWorld system (you can see it in the stack message) that is conflicting with yours: NavLerpSystem, NavSurfaceSystem and NavDestinationSystem
It is them that need to be adjusted with UpdateAfter and UpdateBefore, since I see no point in adding these to the Unity's physics library. Mine is on 0.6.0, maybe you could try updating and see if it is compatible

And again, this is an experience of importing your package as a third party into a project with the latest unity packages. I didn't try cloning the entire repo and seeing whether everything runs, because I bet it does.

**Comment by reeseschultz** [comment]
Since you understand the nature of the conflict, do you want to go ahead and submit a PR fixing the issue?

**Comment by Xroft666** [comment]
I can only offer Dependency.Complete() at this time to you xD

**Comment by sky-higher** [comment]
replace 
buildPhysicsWorld.AddInputDependency(Dependency); 
by
buildPhysicsWorld.AddInputDependencyToComplete(Dependency);

i'm ecs newbie XD

**Comment by reeseschultz** [comment]
I'm updating the nav package such that it fixes this issue (and upgrades to Physics 0.6.0), but it may be slightly slower due to the introduction of sync points. The reason I'm relenting on sync points is that I'm building a new *nav* package devoid of surface parenting, physics, lerping, etc. Essentially, the plan is for it to just be pathfinding, so it will have little-to-no sync points. The user can add them at their own peril. Those who want more control, and want to maximize performance, can use *that* package instead. Check back periodically if it interests you.

**Edit:** See #67.
